bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java

```diff
@@ -847,38 +839,44 @@ protected boolean finalizeStopAfterDisconnect(boolean v) {
     }
 
     /**
-     * Publish a message to the broker with the given QoS and retained flag.
+     * Publish a message to the broker.
      *
      * @param topic The topic
      * @param payload The message payload
-     * @param qos The quality of service for this message
-     * @param retain Set to true to retain the message on the broker
      * @param listener A listener to be notified of success or failure of the delivery.
      */
-    public void publish(String topic, byte[] payload, int qos, boolean retain, MqttActionCallback listener) {
-        MqttAsyncClient localClient = client;
-        if (localClient == null) {
-            listener.onFailure(topic, new MqttException(0));
-            return;
-        }
-        try {
-            IMqttDeliveryToken deliveryToken = localClient.publish(topic, payload, qos, retain, listener,
-                    actionCallback);
-            logger.debug("Publishing message {} to topic '{}'", deliveryToken.getMessageId(), topic);
-        } catch (org.eclipse.paho.client.mqttv3.MqttException e) {
-            listener.onFailure(topic, new MqttException(e));
-        }
+    public void publish(String topic, byte[] payload, MqttActionCallback listener) {
+        publish(topic, payload, null, null, listener);
     }
 
     /**
-     * Publish a message to the broker.
+     * Publish a message to the broker with the given QoS and retained flag.
      *
      * @param topic The topic
      * @param payload The message payload
+     * @param qos The quality of service for this message
+     * @param retain Set to true to retain the message on the broker
      * @param listener A listener to be notified of success or failure of the delivery.
      */
-    public void publish(String topic, byte[] payload, MqttActionCallback listener) {
-        publish(topic, payload, qos, retain, listener);
+    public void publish(String topic, byte[] payload, @Nullable Integer qos, @Nullable Boolean retain, MqttActionCallback listener) {
```

Comment:
Hm. How about forcing the API consumer to set qos and retain? Would that be soo bad?

The argument comments could be "qos The quality of service for this message. **0 is sufficient for most cases**" and "retain Set to true to retain the message on the broker. **In most cases you want to set this to false**".

The api consumer has to specify "null" at the moment anyway, there is not even a typing benefit.

Reply:
Removing the nullable makes it much more difficult to maintain backward compatibility with the QoS and retain parameters on the connection.

Reply:
To stay backwards compatible all you need to offer is the deprecated `publish` method without qos and state. And in that one you would call `publish(topic, payload, this.getQos(), this.getRetain())`.

That's why I do not understand why you need nullable objects here. Maybe I'm overlooking something?

Reply:
QoS is not deprecated on connection level. If we decide to do that, we can keep both as @NonNull. The reason for the @Nullable is to make it possible to set retain with the defualt qos (qos=null) or change the qos and  keep the retained (which is deprecated but should still work). The other option is to have three seperate methods instead of one (qos, retain,qos+retain).

Reply:
Ah ok, your argument is that you may want to set only one of the parameters.
But then again, you can just use the getter of the other parameter to get the default, eg:

`conn.publish(topic, payload, conn.getQos(), the_retain_value)`

Reply:
good point.